src/Umbraco.Core/Services/IdkMap.cs

```diff
@@ -156,16 +163,72 @@ public void ClearCache(Guid key)
             try
             {
                 _locker.EnterWriteLock();
-                int id;
-                if (_key2Id.TryGetValue(key, out id) == false) return;
-                _id2Key.Remove(id);
-                _key2Id.Remove(key);
+                var match = _key2Id.Keys.SingleOrDefault(x => x.Key == key);
+                if (match == null) return;
+                var id = _key2Id[match];
+                var reversedCompositeKey = new Id2KeyCompositeKey {Id = id, UmbracoObjectType = match.UmbracoObjectType};
+                _id2Key.Remove(reversedCompositeKey);
+                _key2Id.Remove(match);
             }
             finally
             {
                 if (_locker.IsWriteLockHeld)
                     _locker.ExitWriteLock();
             }
         }
+
+        internal class Id2KeyCompositeKey
+        {
+            public int Id { get; set; }
```

Comment:
This should be public get, private set so they are immutable and be set on the ctor. In this case this object should probably be a struct instead since it satisfies all of the struct requirements:

* immutable
* has comparison members declared
* is lightweight

To make this immutable with ctor params, you cannot use the fancy {get;private set;} you'l need to go old school and use private readonly fields and return those field values in {get;}